agent/hcp/telemetry_provider.go

```diff
@@ -0,0 +1,199 @@
+package hcp
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"github.com/mitchellh/hashstructure/v2"
+
+	"github.com/hashicorp/consul/agent/hcp/client"
+	"github.com/hashicorp/consul/agent/hcp/telemetry"
+)
+
+var (
+	// internalMetricRefreshFailure is a metric to monitor refresh failures.
+	internalMetricRefreshFailure []string = []string{"hcp", "telemetry_config_provider", "refresh", "failure"}
+	// internalMetricRefreshSuccess is a metric to monitor refresh successes.
+	internalMetricRefreshSuccess []string = []string{"hcp", "telemetry_config_provider", "refresh", "success"}
+)
+
+// Ensure hcpProviderImpl implements telemetry provider interfaces.
+var _ telemetry.ConfigProvider = &hcpProviderImpl{}
+var _ telemetry.EndpointProvider = &hcpProviderImpl{}
+
+// hasher can be implemented to compute the hash of a dynamicConfig.
```

Comment:
@loshz : Drop the entire hashing logic and worry about write lock contention.

From a performance perspective, this won't affect much.